core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -410,32 +413,41 @@ class GroupCoordinator(val brokerId: Int,
           if (generationId < 0) {
             // the group is not relying on Kafka for group management, so allow the commit
             val group = groupManager.addGroup(new GroupMetadata(groupId))
-            doCommitOffsets(group, memberId, generationId, offsetMetadata, responseCallback)
+            doCommitOffsets(group, memberId, generationId, producerId, producerEpoch, offsetMetadata, responseCallback)
           } else {
             // or this is a request coming from an older generation. either way, reject the commit
             responseCallback(offsetMetadata.mapValues(_ => Errors.ILLEGAL_GENERATION))
           }
 
         case Some(group) =>
-          doCommitOffsets(group, memberId, generationId, offsetMetadata, responseCallback)
+          doCommitOffsets(group, memberId, generationId, producerId, producerEpoch, offsetMetadata, responseCallback)
       }
     }
   }
 
+  def handleTxnCompletion(producerId: Long,
+                          topicPartitions: Seq[TopicPartition],
+                          transactionResult: TransactionResult) {
+    val offsetPartitions = topicPartitions.filter(_.topic() == Topic.GroupMetadataTopicName).map(_.partition).toSet
+    groupManager.handleTxnCompletion(producerId, offsetPartitions, transactionResult == TransactionResult.COMMIT)
+  }
+
   private def doCommitOffsets(group: GroupMetadata,
                               memberId: String,
                               generationId: Int,
+                              producerId: Long,
+                              producerEpoch: Short,
                               offsetMetadata: immutable.Map[TopicPartition, OffsetAndMetadata],
                               responseCallback: immutable.Map[TopicPartition, Errors] => Unit) {
     var delayedOffsetStore: Option[DelayedStore] = None
-
     group synchronized {
       if (group.is(Dead)) {
         responseCallback(offsetMetadata.mapValues(_ => Errors.UNKNOWN_MEMBER_ID))
-      } else if (generationId < 0 && group.is(Empty)) {
+      } else if ((generationId < 0 && group.is(Empty)) || (producerId != RecordBatch.NO_PRODUCER_ID)) {
```

Comment:
Oh actually, that's wrong. We can fail the `TxnOffsetCommit`. I was thinking this was part of the roll forward logic. So then I guess the question is whether it's worthwhile keeping some additional state to be able to tell whether the group has done consumer offset commits for the current generation. If it has, perhaps we should send back an error.